util/tcpinfo_linux.go

```diff
@@ -0,0 +1,129 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package util
+
+import (
+	"syscall"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+func getConnTCPInfo(raw syscall.RawConn) (*TCPInfo, error) {
+	var info linuxTCPInfo
+	size := unsafe.Sizeof(info)
+
+	var errno syscall.Errno
+	err := raw.Control(func(fd uintptr) {
+		_, _, errno = unix.Syscall6(unix.SYS_GETSOCKOPT, fd, unix.IPPROTO_TCP, unix.TCP_INFO,
+			uintptr(unsafe.Pointer(&info)), uintptr(unsafe.Pointer(&size)), 0)
+	})
+	if err != nil {
+		return nil, err
+	}
+	if errno != 0 {
+		return nil, errno
+	}
+	if info == (linuxTCPInfo{}) {
+		return nil, ErrNoTCPInfo
+	}
+	return &TCPInfo{
+		RTT:        info.rtt,
+		RTTVar:     info.rttvar,
+		RTTMin:     info.min_rtt,
+		SndMSS:     info.snd_mss,
+		RcvMSS:     info.rcv_mss,
+		SndCwnd:    info.snd_cwnd, // Send congestion window
+		RcvWnd:     info.snd_wnd,  // "tp->snd_wnd, the receive window that the receiver has advertised to the sender."
+		Rate:       info.delivery_rate,
+		AppLimited: bool((info.app_limited >> 7) != 0), // get first bit
+	}, nil
+}
+
+// linuxTCPInfo is based on linux include/uapi/linux/tcp.h struct tcp_info
+//revive:disable:var-naming
+type linuxTCPInfo struct {
```

Comment:
My first inclination was to embed the Go version of this struct instead, and then add the new fields. But I suppose that would break if/when Go _does_ add the fields.

Still, I'm half tempted to say to embed it anyway, and add unit test to fail if the size of the Go struct changes.

Reply:
Makes sense, I had assumed they only added to the end, rather than specifying some bytes in an early word.